### DISCUSSION TITLE
WSI DataLoader

### DISCUSSION THREAD

**Discussion by JHancox** [Project-MONAI/MONAI#3608] [Ideas]
MONAI Pathology might serve the community well by producing a DataLoader that can be pointed at a directory of WSIs and it can take care of the loading of the pixels and labels using the most efficient tools and techniques available. This would remove a lot of repetitive code that is often creates bottlenecks in the training or inference pipeline. 

Since MONAI is based on PyTorch, this class would be subclassed from the standard DataLoader and would be compatible with other PyTorch techniques such as DistributedDataParallel, supporting multiprocessing. The actual loading could be undertaken by the most optimal loader for the file format, with a preference for leveraging the performance provided by cuCIM, but the ability to use the best component for the job.

There are a number of possible features this could support such as:
- Dynamic thresholding and batch preparation
- Randomisation of file and patch order
- Automatic label/image collation
- Automatic epoch/batch size handling
- Multi-GPU support

Feedback and ideas for features would be appreciated

**Comment by cooperlab** [comment]
Here are some thoughts on features:
- Control over rescaling behavior if desired magnification not available (anti-aliasing option if possible)
- Using a binary mask (possibly lower res) to filter tiles
- Accept user-defined tile filtering function that consumes tile and produces bool
- Weighting tiles or weighted sampling from a continuous-valued mask
- Interleaving tiles from multiple slides (likely supported by PyTorch)
- Jitter tile positions for training. Easy to implement by increasing tile size followed by random crop to desired size.

@Leengit and I are developing a similar TensorFlow loader to create queued+prefetched tile streams https://github.com/DigitalSlideArchive/histomics_stream. This uses largeimage to interface with WSI files but we are looking at more performant alternatives like CuCIM or a DALI plugin. I think it provides some nice options for mapping between available and desired magnifications.

**Comment by OmarAshkar** [comment]
+1 for this concept. Repetitive code to load WSI is time consuming.

**Comment by tangy5** [comment]
+1 Do we already have this? I'm seeing the prior works, data loading proccess for WSI is very inefficient, 90% of time is used for reading images, resulting low GPU utilities during training.

**Comment by Geeks-Sid** [comment]
Any updates on this issue from the team?

**Comment by gafaua** [comment]
Hello, I don't know if I can post this here, but I've developed a lightweight library for a personal project for easy handling of WSIs in training loops. I think it covers many of the topics discussed in this issue.

https://github.com/gafaua/wsi-dataloader

(Please remove this comment if this is not allowed)
